cci_cameroon/analysis/clustering_rumours/rumour_clustering_model_development.py

```diff
@@ -0,0 +1,1048 @@
+id_to_category = dict(model_data[["category_id", "code"]].values)
+model_data = (
+    model_data.groupby("comment")["category_id"].apply(list).to_frame().reset_index()
+)
+
+# %%
+model_data = model_data.reset_index()
+
+# %%
+model_data.shape
+
+# %%
+english_df = data_df[data_df.language == "en"].copy()
+spanish_df = data_df[data_df.language == "es"].copy()
+data_df = data_df[~data_df.language.isin(["en", "es"])].copy()
+
+# %%
+english_df["comment"] = english_df.comment.apply(
+    translator.translate, src="en", dest="fr"
+).apply(getattr, args=("text",))
+spanish_df["comment"] = spanish_df.comment.apply(
+    translator.translate, src="es", dest="fr"
+).apply(getattr, args=("text",))
+
+# %%
+data = pd.concat([data_df, english_df, spanish_df], ignore_index=True)
+
+# %%
+data.drop("language", axis=1, inplace=True)
+
+# %%
+data.shape
+
+# %%
+data["category_id"] = data.first_code.factorize()[0]
+
+# %%
+# checking the distribution of the codes in the data
+data.groupby("first_code").comment.count().plot(kind="bar")
+
+# %%
+data = data.drop_duplicates(["comment"])
+data.reset_index(inplace=True)
+data.shape
+
+# %%
+# Converting the codes into integers to act as ground truth
+data["cluster"] = data.code.factorize()[0]
+
+# %%
+# using the french_semantic model for word embedding
+model = SentenceTransformer("Sahajtomar/french_semantic")
+# model = SentenceTransformer("distiluse-base-multilingual-cased-v1")
+
+# %%
+data[
+    data.category_id == 2
+].shape  # 127 cluster 1(0),137 cluster 2 (1), 136 cluster 3 (2)
+
+# %% [markdown]
+# ## Using a smaller sample of the data (~30 records) to ease inspection of results
+
+# %%
+# creating a small subset of the data made up of the different categories
+set1 = data[data.category_id == 0][["comment", "category_id"]][:10]
+set2 = data[data.category_id == 1][["comment", "category_id"]][:10]
+set3 = data[data.category_id == 2][["comment", "category_id"]][:10]
+to_use2 = (
+    pd.concat([set1, set2, set3]).drop_duplicates().reset_index().drop("index", axis=1)
+)
+# to_use2 = to_use.reset_index().drop("index",axis=1)
+
+# %%
+# create word embeddings using the transformer model
+sentence_embeddings2 = model.encode(to_use2.comment)
+
+# %%
+indp2 = faiss.index_factory(
+    sentence_embeddings2.shape[1], "Flat", faiss.METRIC_INNER_PRODUCT
+)
+faiss.normalize_L2(sentence_embeddings2)  # to be used for inner product computation
+# setting the number of neighbors to consider for graph connectivity
+neighbors2 = 10
+indp2.train(sentence_embeddings2)
+indp2.add(sentence_embeddings2)
+similarity_distance, sim_positions = indp2.search(
+    sentence_embeddings2, neighbors2  # consider only the first n elements
+)
+
+# %%
+edges, weighted_edges = generate_edges(
+    to_use2, "comment", sim_positions, similarity_distance
+)
+
+# %%
+nodes = list(set(to_use2.comment))
+
+# %%
+NG = nx.Graph()
+NG.add_nodes_from(nodes)
+NG.add_edges_from(edges)
+NG.add_weighted_edges_from(weighted_edges)
+
+# %%
+nx.draw(NG)
```

Comment:
Looks really nice! The next step will be us saving these to ouput/figures but I think we can make these calls when we have all files in the model pipeline mother branch.